### PR TITLE
Major overhaul, primarily configuration support

### PR DESCRIPTION
**Pull request by martin-oom**
* add support for creating database and collections and deleting database (IDatabaseOperations)
 * removes requirement for using DatabaseIdAttribute, fetched from connection string or defaults to database class name
 * removes requirement for using CollectionIdAttribute, defaults to property name
 * moves internal declarations to "Internals" directory
 * includes XML-documentation
 * adds code analyzers (fxcop)
 * adds functional tests
 * adds example application (TodoApi)
 * Removes old static class TermAnalyzer, replaced by configurable Analyzer
 * Removes old entity extension GetAllTextFieldValues, replaced by ITextExtractor
 * Adds DatabaseModel and CollectionModel for configuration and metadata
 * Adds generic CosmosDatabase with configuration support (OnModelCreating)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -28,14 +28,14 @@ public class MyEntity : ICosmosEntity
...
+[DatabaseId("MyDatabase")] // name of the cosmos database, optional
 public class MyDatabase : CosmosDatabase
 {
     public MyDatabase(DatabaseOptions<MyDatabase> options)
         : base(options)
     { }
 
-    [CollectionId("MyEntities")] // name of the cosmos collection, required
+    [CollectionId("MyEntities")] // name of the cosmos collection, optional
     public ICosmosDbCollection<MyEntity> MyEntities { get; private set; }
```
Comment (alveflo) [inline README.md:39]: So if I don't specify the name of the cosmos collection the name will be `nameof(MyEntities)`?

File: Spinit.CosmosDb.FunctionalTests/Infrastructure/TestCaseByAttributeOrderer.cs [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+using Xunit.Sdk;
+
+namespace Spinit.CosmosDb.FunctionalTests.Infrastructure
+{
+    public class TestCaseByAttributeOrderer : ITestCaseOrderer
+    {
+        public IEnumerable<TTestCase> OrderTestCases<TTestCase>(IEnumerable<TTestCase> testCases)
+            where TTestCase : ITestCase
+        {
+            // this looks like crap but has to be done due to strange implementation by Xunit that seems to use some sort of dynamic proxy...
```
Comment (alveflo) [inline Spinit.CosmosDb.FunctionalTests/Infrastructure/TestCaseByAttributeOrderer.cs:13]: 😄

File: Spinit.CosmosDb/Analysis/AnalyzeContext.cs [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+{
+    public enum AnalyzeContext
+    {
+        /// <summary>
+        /// Analyze tokens for an entity used when indexing/upserting
+        /// </summary>
+        Entity = 1,
+
+        /// <summary>
+        /// Analyze tokens for a search quey
```
Comment (alveflo) [inline Spinit.CosmosDb/Analysis/AnalyzeContext.cs]: Typo

File: README.md (hunk 1 above)
Comment (martin-oom) [reply README.md:39]: Yes, the name of the collection is set in this order:
1. from collectionmodel, eg only when using generic CosmosDatabas (with custom configuration support). 
2. from attribute (CollectionIdAttribute)
3. from "default" eg the name of the property

File: Spinit.CosmosDb/Analysis/AnalyzeContext.cs (hunk 3 above)
Comment (martin-oom) [reply Spinit.CosmosDb/Analysis/AnalyzeContext.cs]: Wow, thanks Hawkeye! :-)

**Pull request merged**